### PR TITLE
CDAP-15628 Added basic query parsing API functionality and test cases

### PR DESCRIPTION
**Pull request by grammarstudies**
https://builds.cask.co/browse/CDAP-DUT7001

### CODE REVIEW

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import io.cdap.cdap.common.metadata.QueryTerm.Qualifier;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Pattern;
+
+/**
+ * A thread-safe class that provides helper methods for metadata search string interpretation,
+ * and defines search syntax for qualifying information (e.g. required terms) {@link QueryTerm.Qualifier}.
+ * By default, search items without an operator are considered optional.
+ */
```
Comment (anew) [inline cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java:28]: The javadoc for this class (or for the parse() method should contain a precise description of the syntax it parses:
- what characters are considered as whitespace/separators?
- what is the syntax for a required term
- is it case-sensitive / are terms lower-cased?

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+public class QueryParser {
+  private static final Pattern SPACE_SEPARATOR_PATTERN = Pattern.compile("\\s+");
+  private static final char REQUIRED_OPERATOR = '+';
+
+  /**
+   * Organizes and separates a raw, space-separated search string
+   * into multiple {@link QueryTerm} objects.
+   * e.g. "tag1 tag2 tag3" is converted to a list of optional QueryTerms containing the strings
+   * "tag1", "tag2", and "tag3" respectively.
+   * @param query the raw search string
```
Comment (anew) [inline cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java:49]: Nit: there should be an empty line before the @param

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    for (String term : Splitter.on(SPACE_SEPARATOR_PATTERN)
+        .omitEmptyStrings().trimResults().split(query)) {
+      queryTerms.add(singleParse(term));
+    }
+    return queryTerms;
+  }
+
+  private static QueryTerm singleParse(String term) {
+    if (term.charAt(0) == REQUIRED_OPERATOR) {
+      if (term.length() == 1) {
```
Comment (anew) [inline cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java]: this does exactly the same as the else branch? What do you think it means if there is a single "+"? Can anything match a required empty term?

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryTerm.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+    this.qualifier = qualifier;
+  }
+
+  public String getTerm() {
+    return term;
+  }
+  public Qualifier getQualifier() {
+    return qualifier;
+  }
+  public boolean equals(QueryTerm q) {
```
Comment (anew) [inline cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryTerm.java]: The standard way to define equals is as 
```
@Override 
public boolean equals(Object other)
```
And if you implement equals, you must also implement hashCode.

File: cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+  public void testBasicParse() {
+    String inputQuery = "tag1 tag2 +tag3 tag4+";
+    List<QueryTerm> targetOutputQueryTerms = new ArrayList<>();
+    targetOutputQueryTerms.add(new QueryTerm("tag1", Qualifier.OPTIONAL));
+    targetOutputQueryTerms.add(new QueryTerm("tag2", Qualifier.OPTIONAL));
+    targetOutputQueryTerms.add(new QueryTerm("tag3", Qualifier.REQUIRED));
+    targetOutputQueryTerms.add(new QueryTerm("tag4+", Qualifier.OPTIONAL));
+    List<QueryTerm> actualOutputQueryTerms = QueryParser.parse(inputQuery);
+
+    for (int i = 0; i < 4; i++) {
```
Comment (anew) [inline cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java]: This would miss the case where the actual terms have a 5th element. I think it is more concise to use 
```
Assert.assertEquals(targetOutputQueryTerms, actualOutputQueryTerms);
```

File: cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+import org.junit.Assert;
+import org.junit.Test;
+import java.util.ArrayList;
+import java.util.List;
+
+public class QueryParserTest {
+  @Test
+  public void testBasicParse() {
+    String inputQuery = "tag1 tag2 +tag3 tag4+";
+    List<QueryTerm> targetOutputQueryTerms = new ArrayList<>();
```
Comment (anew) [inline cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java]: You could use an ImmtuableList.of(term1, term2, ...)

File: cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+  @Test
+  public void testUnusualQueries() {
+    String inputQuery1 = "";
+    String inputQuery2 = "        space";
+    String inputQuery3 = "+ plus";
+
+    Assert.assertTrue(QueryParser.parse(inputQuery1).isEmpty());
+    Assert.assertTrue(QueryParser.parse(inputQuery2).get(0).equals(new QueryTerm("space", Qualifier.OPTIONAL)));
+    Assert.assertTrue(QueryParser.parse(inputQuery3).get(0).equals(new QueryTerm("", Qualifier.REQUIRED)));
```
Comment (anew) [inline cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java]: Is this really the behavior that we want?

File: cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+package io.cdap.cdap.common.metadata;
+
+import io.cdap.cdap.common.metadata.QueryTerm.Qualifier;
+import org.junit.Assert;
+import org.junit.Test;
+import java.util.ArrayList;
+import java.util.List;
+
+public class QueryParserTest {
```
Comment (anew) [inline cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java:25]: to be complete, you should have more test cases and cover edge cases

- single term
- +*
- other white space than blanks
- empty query

etc.
One way to do that could be to define an array or a map of all test cases string -> expected result. And then have one test method that iterates over all of these entries and tests them.

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java (hunk 1 above)
Comment (grammarstudies) [reply cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java:28]: As discussed, the javadoc now points readers to the SPACE_SEPARATOR_PATTERN field and java.util.regex.Pattern class for information on separator characters.
It points readers to the REQUIRED_OPERATOR field to explain the required term syntax.
It mentions that terms keep the same case when parsed, only losing any applicable operators.

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java (hunk 2 above)
Comment (grammarstudies) [reply cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java:49]: Fixed.

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java (hunk 3 above)
Comment (grammarstudies) [reply cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java]: The else branch redundancy has been fixed.
The meaning of a single "+" has been updated. Now, an operator that does not precede a search term will be considered to _be_ the search term—the "+" will be searched for, and not counted as an operator.

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryTerm.java (hunk 4 above)
Comment (grammarstudies) [reply cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryTerm.java]: Fixed.

File: cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java (hunk 6 above)
Comment (grammarstudies) [reply cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java]: Fixed.

File: cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java (hunk 5 above)
Comment (grammarstudies) [reply cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java]: Fixed.

File: cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java (hunk 7 above)
Comment (grammarstudies) [reply cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java]: The behavior has been improved as per the comments on the QueryParser's parsing method.

File: cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java (hunk 8 above)
Comment (grammarstudies) [reply cdap-common/src/test/java/io/cdap/cdap/common/metadata/QueryParserTest.java:25]: More test cases have been added to cover these.
Regarding the "+*" case, since the interpretation of a wildcard is left to the individual users of QueryParser, would it be better to handle that case within the users' tests?

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java (hunk 1 above)
Comment (anew) [reply cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryParser.java:28]: nice!

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryTerm.java [hunk 9]
```diff
@@ -34,18 +36,36 @@
   }
   private Qualifier qualifier;
 
+  /**
+   * Constructs a QueryTerm using the search term and its qualifying information.
+   * @param term the search term
```
Comment (anew) [inline cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryTerm.java:42]: Nit: missing newline

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryTerm.java [hunk 10]
```diff
@@ -62,7 +62,12 @@ public Qualifier getQualifier() {
 
   @Override
   public boolean equals(Object o) {
-    return this.getClass() == o.getClass() && this.hashCode() == o.hashCode();
+    if (o == this) return true;
```
Comment (anew) [inline cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryTerm.java]: This will not pass checkstyle, because you are missing the { .. }

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryTerm.java (hunk 9 above)
Comment (grammarstudies) [reply cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryTerm.java:42]: Addressed in latest commit.

File: cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryTerm.java (hunk 10 above)
Comment (grammarstudies) [reply cdap-common/src/main/java/io/cdap/cdap/common/metadata/QueryTerm.java]: Addressed in latest commit.

**Review by anew — APPROVED** [review]
lgtm
